rfd/0038-resource-encryption.md

```diff
@@ -0,0 +1,76 @@
+
+AES-CBC-HMACSHA256 was chosen because it is a popular AEAD construct supported by various libraries including Google Tink which can be built to rely on BoringCrypto primitives.
+
+The asymmetric cipher that will be used is RSA-2048 as it is widely compliant and has trusted implementations.
+
+### Local and remote
+
+Currently, encrypting resources at-rest on remote services is of the highest priority.
+When these need to be accessed by an auth server on request of a node or these resources must be fetched from the remote services and decrypted and cached locally for at least the duration of the access.
+
+Because this is a form of client side encryption we do not rely on the service itself and thus we should strive to support encryption on all remote storage services Teleport supports where possible including both AWS and GCP.
+
+### Key management
+
+When using symmetric encryption algorithms to encrypt and decrypt data on auth servers a strategy is needed for managing these keys and making sure all auth servers in a high availability deployment can decrypt data from all other auth servers.
+
+Broadly, there are two options available:
+
+1. Sharing a single encryption key across all authentication servers.
+2. Taking a card from the AWS Encryption SDK playbook and using a two layer key system.
+
+Option 1 has multiple significant drawbacks which make it unsuitable for production use. The major issue is that all data is fundamentally tied to the only layer of security meaning that if a malicous party potentially acquired access hold to the key, all data would have to be reincrypted. This makes key rotation prohibitive. Another major issue is that of prolonged key reuse which contradicts industry best practices.
+
+Option 2 is doing something very similar to what the AWS Encryption SDK does. Ideally using that library would be ideal since it is a very nice encryption toolkit but unfortunately it isn't available for Go. The strategy involves using unique ephemeral data keys for each encrypted object (rotated when reincrypted/modified) that are encrypted and bundled with each encrypted object in a defined format. A central PEM encoded master keypair stored in a new resource is used to encrypt these per object keys.
+
+Option 2 here seems like a far more robust solution for production use and that is thus what we should employ.
+
+### Encryption format
+
+For facilitating key rotation the encrypted objects should be stored in a seperate file from it's data key bundle. This is due to modification restricts on objects imposed by services such as AWS S3.
+
+All encrypted objects should be suffixed by a `.enc` extension to communicate that they are encrypted. For simplicity no other data should be bundled in the same file.
+
+The file that stores the encrypted data key should be a file with the name `$objectName.keys` stored next to the encrypted object.
```

Comment:
keys plural? How will we map multiple DEKs to a single object? Also, Will this file just be a 64-bit version header + key bits, or will there be a higher-level structure to it?

Reply:
Ah good typo, that should be .key. Regarding the version I am unsure of the best way to do this. I had another alternative pop up in my head that we use a JSON document for the key instead containing the encrypted key data and versions for both the key and the data object. That makes things slightly less convoluted.

Reply:
I went ahead and changed the format slightly to remove the raw header as it introduces some uneeded complexity. Please have another look at this section.